reco_utils/dataset/pandas_df_utils.py

```diff
@@ -166,3 +167,97 @@ def _convert(field, feature, field_index, field_feature_index_dict):
+    binarized feedback, i.e., 1 and 0 indicate positive and negative feedback, 
+    respectively. 
+
+    Negative sampling is used in the literature frequently to generate negative samples 
+    from a user-item interaction data.
+    See for example the neural collaborative filtering paper 
+    https://www.comp.nus.edu.sg/~xiangnan/papers/ncf.pdf
+    
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+            'userID': [1, 2, 3],
+            'itemID': [1, 2, 3],
+            'rating': [5, 5, 5]
+        })
+        >>> df_neg_sampled = negative_feedback_sampler(
+            df, col_user='userID', col_item='itemID', number_neg_per_user=1
+        )
+        >>> df_neg_sampled
+        userID  itemID  feedback
+        1   1   1
+        1   2   0
+        2   2   1
+        2   1   0
+        3   3   1
+        3   1   0
+
+    Args:
+        df (pandas.DataFrame): input data that contains user-item tuples.
+        col_user (str): user id column name.
+        col_item (str): item id column name.
+        col_label (str): label column name. It is used for the generated columns where labels
+        of positive and negative feedback, i.e., 1 and 0, respectively, in the output dataframe.
+        number_neg_per_user (int): number of negative feedback sampled for each user. 
+        seed (int): seed for the random state of the sampling function.
+
+    Returns:
+        pandas.DataFrame: data with negative feedback 
+    """
+    columns = df.columns
+    if col_user not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_user))
+
+    if col_item not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_item))
+
+    # Get all of the users and items.
+    users = df[col_user].unique()
+    items = df[col_item].unique()
+
+    # Create a dataframe for all user-item pairs
+    user_item_tuples = [(row[col_user], row[col_item]) for _, row in df.iterrows()]
+
+    # Generate feedback for both positive and negative
+    user_item_feedback = [
+        (user, item, 1 if (user, item) in user_item_tuples else 0) for user in users for item in items
+    ]
+    df_all = pd.DataFrame(user_item_feedback, columns=[col_user, col_item, col_label])
+
+    # Take all positive feedback
+    df_pos = df_all[df_all[col_label] == 1]
+
+    # Sample # negative feedback for each user
+    # If the total negative feedback for a user is less than the sampling size, all of the 
+    # negative feedback will be generated.
+    df_neg = df_all[df_all[col_label] == 0]
+    df_neg_sample = (
+        df_neg
+        .groupby(col_user)
+        .apply(lambda x: x.sample(
+            n=min(number_neg_per_user, len(x)),
+            random_state=seed, 
+            replace=False))
+        .reset_index(drop=True)
+    )
+
+    # Concatenate the positive and negative feedback dataframes
+    df_sampled = (
```

Comment:
Correct me if I'm thinking wrong...but I have feeling that we maybe able to reuse stratified sampling function or chrono-splitter for this task... e.g. if you set 1 for positive and 0 as negative and regards them as 'timestamp' and run w/ chrono-splitter.

Reply:
Hmmm this is really an interesting and great idea. I am not sure if it can be nicely coped with the existing splitter method - as you said random sampling may be an issue and sample the negative feedback items w.r.t the positive samples can also be something tricky. In that case, perhaps reusing the splitter may introduce even more overhead to the code 😅

Reply:
@loomlike I will merge now and then if there is anything to refactor I'll revisit the code.